pkg/challenges/challenges.go

```diff
@@ -456,7 +456,7 @@ func isSpiffeIDAllowed(host, spiffeID string) bool {
 	if u.Hostname() == host {
 		return true
 	}
-	return strings.Contains(u.Hostname(), "."+host)
+	return strings.HasSuffix(u.Hostname(), "."+host)
```

Comment:
I actually think we should be stricter than this and not allow subdomains. Otherwise, we loosen the definition of a SPIFFE trust domain. A SPIFFE ID is composed of a trust domain and a workload ID. The `1.example.com` trust domain differs from the `2.example.com` trust domain. I'd prefer we be strict and require exact trust domain name matching.

Otherwise, we should be doing something similar to the URI/username flows, and require specification of both the issuer and trust domain up front, require those to match as per `validateAllowedDomain`, and check that the trust domain matches what's in the configuration. This is useful if you want to host a key on a subdomain (`discovery.example.com`) but your trust domain is `example.com` - Prove ownership over both up front. Both `strings.Contains` and `strings.HasSuffix` don't handle this case.

For now, I'd prefer we restrict the match to be exact, especially since SPIFFE OIDC providers aren't in use - We can add support for the OIDC issuer on a different subdomain later.

Reply:
I'm hesitant about potentially breaking something. There appears to be a SPIFFE (actually two, but one is Dans) in prod https://github.com/sigstore/fulcio/blob/fdaedd62f94bb94d656f6fc64c007cbff335fe80/config/fulcio-config.yaml#L25-L29 and there may be more hosted by others.

The OIDC discovery endpoint isn't part of the SPIFFE spec at all as far as I've seen though there is this helper OIDC discover server if you're running the spire implementation https://github.com/spiffe/spire/blob/main/support/oidc-discovery-provider/README.md so I'm not 100% what the allowed behaviour is here. The cross domain behaviour that this PR blocks has a very obvious _yikes_ energy to it though so I'm glad we're going to stop that.

Reply:
Asked folks in spiffe slack if this behaviour is actually specified somewhere or what ever https://spiffe.slack.com/archives/CBNCC2V17/p1652734651335059

Reply:
The correct behavior should be to do exact trust domain matching. By allowing subdomain matching, we're not actually doing trust domain checks.

An example of the threat:
* X hosts an OIDC issuer on `example.com`.
* X creates a trust domain for A, `cocacola.example.com` 
* X creates a trust domain for B, `pepsi.example.com`
* OIDC tokens are issued for a workload in each trust domain. Same OIDC issuer, but different trust domain.
* X registers `example.com` as an OIDC issuer with Fulcio, planning to only have A fetch code-signing certs for `cocacola.example.com`
* B, only having control over `pepsi.example.com`, realizes this subdomain matching, and requests a code-signing cert for `cocacola.example.com`.

OIDC and SPIFFE should be treated entirely separately. If you want to host an OIDC discovery endpoint on a different domain than the SPIFFE trust domain, that could be allowed too! We don't even need to enforce a partial match between the two. But we *must* enforce that the trust domains match.

`allow.pub` is inactive I believe, though we can contact the owner to confirm. Changing behavior because of security is an acceptable breaking change.

Reply:
To your question on the Slack channel, no, there's nothing in the spec that mandates the OIDC issuer and trust domain match. That's just if we want to be more restrictive, it limits some risk.

The issue with the current implementation is independent of OIDC, it's that we are too lax. A trust domain has no notion of subdomains, it's just a string.

Reply:
Ah right so this code, which checks for a relationship between the OIDC issuer URL and the trust domain of the SPIFFE ID needs to be throw alllll the way out right? That's just a bogus idea. We need to configure the trust domain in the Fulcio config and check exact match against the SPIFFE ID / subject in OIDC tokens regardless of what the OIDC issuer domain is right?

😬 This is more broken than originally anticipated 😆

Reply:
I'll add the config code for that. Happy to make that a breaking config change as well. Like you said, its got a kind of nasty security assumption baked into right now

Reply:
> We need to configure the trust domain in the Fulcio config and check exact match against the SPIFFE ID / subject in OIDC tokens regardless of what the OIDC issuer domain is right?

Yes, correct! Similar to how we handle [uri](https://github.com/sigstore/fulcio/blob/8e2f0e1815c147e17b06046a6d2678a9227e12d0/pkg/challenges/challenges.go#L276), we should check that the configured trust domain exactly matches the token's trust domain.

I'd like to also require that the OIDC issuer configuration partially matches the trust domain configuration like [here](https://github.com/sigstore/fulcio/blob/8e2f0e1815c147e17b06046a6d2678a9227e12d0/pkg/challenges/challenges.go#L345-L357). While this isn't a requirement, and we can loosen this restriction in the future, it limits some risk associated with misconfiguration.

Reply:
On it! Thanks for clarifying whats up here!

Reply:
Thanks for the ping in slack, @nsmith5. As already in this conversation, the SPIFFE specifications don't have any defined behavior for OIDC-style federation. The OIDC Discovery Provider provided by the SPIRE project provides a poor persons version of federation for SPIFFE-unaware systems that already know how to verify a JWT using an OIDC Discovery document and the linked JWKS.

Like [SPIFFE Federation](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Federation.m) however, the JWKS pulled from the `jwks_uri` (or the bundle endpoint) should only contain the public keys of the signing authorities for an _individual_ trust domain. As @haydentherapper mentioned, serving the authorities from multiple trust domains from a single endpoint allows one trust domain to impersonate identities from the other trust domain. Maintaining a very specific binding between a trust domain and a bundle endpoint is a paramount [security consideration](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Federation.md#73-preserving-the-trust-domain-bundle-binding).

Reply:
I think its worth considering (in the future) also to make fulcio SPIFFE aware and implement the SPIFFE interfaces, shouldn't be that big a jump considering the work that you're doing splitting up the oidc domains - @haydentherapper .

Can't remember which of the sigstore projects, (maybe cosign?), but SPIFFE support today isn't actually SPIFFE aware (i.e. retrieval of endpoint bundles), just parsing of SPIFFE common name credentials.

Reply:
Lemme split this off into an issue.

Fulcio doesn't retrieve trust bundles, it's just integrated with OIDC. The SPIFFE+Cosign integration is a provider (https://github.com/sigstore/cosign/blob/main/pkg/providers/spiffe/spiffe.go) that fetches a JWT SVID, I think there's an assumption that whatever signs that SVID exposes its verification keys over OIDC too.